engine/googletest_helpers.cpp

```diff
@@ -0,0 +1,344 @@
+static void
+test_timeout(void)
+{
+    std::cout << incomplete_test;
+
+    ::sleep(10);
+    const fs::path control_dir = fs::path(utils::getenv("CONTROL_DIR").get());
+    std::ofstream file((control_dir / "cookie").c_str());
+    if (!file)
+        fail("Failed to create the control cookie");
+    file.close();
+}
+
+
+}  // anonymous namespace
+
+
+/// Prints out program usage and exits with a non-zero exit code.
+static void
+usage(const char* argv0) {
+    std::cout << "usage: " << argv0 << " "
+              << "[--gtest_color=*] "
+              << "[--gtest_filter=POSITIVE_PATTERNS] "
+              << "[--gtest_list_tests]"
+              << "\n\n"
+              << "This program mocks a googletest test program."
+              << "\n";
+    std::exit(EXIT_FAILURE);
+}
+
+/// Entry point to the test program.
+///
+/// The caller can select which test scenario to run by modifying the program's
+/// basename on disk (either by a copy or by a hard link).
+///
+/// \todo It may be worth to split this binary into separate, smaller binaries,
+/// one for every "test scenario".  We use this program as a dispatcher for
+/// different "main"s, the only reason being to keep the amount of helper test
+/// programs to a minimum.  However, putting this each function in its own
+/// binary could simplify many other things.
+///
+/// \param argc The number of CLI arguments.
+/// \param argv The CLI arguments themselves.  These are not used because
+///     Kyua will not pass any arguments to the plain test program.
+int
+main(int argc, char** argv)
+{
+    using scenario_fn_t = void (*)(void);
+
+    std::map<std::string, scenario_fn_t> scenarios;
+    std::string testcase;
+
+    const char *gtest_color_flag = "--gtest_color=";
+    const char *gtest_filter_flag = "--gtest_filter=";
+    char *argv0 = argv[0];
+
+    bool list_tests = false;
+
+    scenarios["check_configuration_variables"] =
+        test_check_configuration_variables;
+    scenarios["crash"] = test_crash;
+    scenarios["fail"] = test_fail;
+    scenarios["pass"] = test_pass;
+    scenarios["pass_but_exit_failure"] = test_pass_but_exit_failure;
+    scenarios["timeout"] = test_timeout;
+
+    /// NB: this avoids getopt_long*, because its behavior isn't portable.
+    int i;
+    for (i = 1; i < argc; i++) {
+        if (strcmp("--gtest_list_tests", argv[i]) == 0) {
+            list_tests = true;
+        }
+        /// Ignore `--gtest_color` arguments.
+        else if (strncmp(gtest_color_flag, argv[i],
+                         strlen(gtest_color_flag)) == 0) {
+            continue;
+        }
+        /// Try looking for the test name with `--gtest_filter`.
+        else {
+            std::string filter_flag_match =
+                std::string(gtest_filter_flag) + test_suite;
+            std::string arg = std::string(argv[i]);
+            if (arg.find_first_of(filter_flag_match) == arg.npos) {
+                usage(argv0);
```

Comment:
Something apparently went wrong. But what?

http://julio.meroh.net/2013/08/cli-design-requesting-and-offering-help.html#only-show-help-when-explicitly-told-to-do-so (my opinion only, but this is what guides Kyua behavior).